### ISSUE TITLE
[arcface_torch]What does fp16 stand for ? in arcface_pytorch

### ISSUE THREAD

**Issue by bokey007** [deepinsight/insightface#1456]
What does fp16 stand for ? in arcface_pytorch

is it for fartial fully connected ? If yes i can see optimiser for partial full connected updates weight even if config.fp16 = False

Any help would be appriciated.
Thanks.

**Comment by anxiangsir** [comment]
fp16 means MIXED PRECISION TRAINING.
see https://arxiv.org/pdf/1710.03740.pdf

Mixed precision training is faster than floatpoint32 without losing model accuracy or having to modify hyperparameters.

**Comment by bokey007** [comment]
Thank you anxiangsir. 

one more question
Are you using Partial fully connected layee in arcface pytorch by default, asking  since i wanted to run without it.

**Comment by anxiangsir** [comment]
Default is model parallel, because sample rate is 1.0, when sample rate changed, partial fc will work.

**Comment by bokey007** [comment]
ohh ok thanks.

I am new to deep learning.

I dont understand the need to create two different optimisers as follows:

    opt_backbone = torch.optim.SGD(
        params=[{'params': backbone.parameters()}],
        lr=cfg.lr / 512 * cfg.batch_size * world_size,
        momentum=0.9, weight_decay=cfg.weight_decay)

    opt_pfc = torch.optim.SGD(
        params=[{'params': module_partial_fc.parameters()}],
        lr=cfg.lr / 512 * cfg.batch_size * world_size,
        momentum=0.9, weight_decay=cfg.weight_decay)

It was my understanding that we can create only one optimiser for one optimisation graph.

**Comment by alicera** [comment]
arcface_pytorch model have the rank0 - rank7 and mom
what does it mean?

**Comment by bokey007** [comment]
seems like they are using 8 GPU for training hence rank 0 to rank 7
what didi u mean by mom?

**Comment by alicera** [comment]
![image](https://user-images.githubusercontent.com/59388989/114998547-a6e14000-9ed3-11eb-9785-f3bab7042eb0.png)
the model name have the mom

**Comment by bokey007** [comment]
seems something related to momentum_buffer. I dont have much idea bout this.

Why dont you create seaparate issue for this.

**Comment by bokey007** [comment]
Hi Alicera, 

I am new to deep learning, so I could be worng.

But as per my understanding; when you want to coninue traning in steps you need to save both weights and  the history of gradiests (here refeered as momentum buffer)... refere SGD with momentum for more details.

and since the wight matrix of last fully connected layer is distributed on 8 GPU--> you can see 16 files. 2 files for each gpu ... one for wieight and one for momentum buffer. ,...---->refer to partial_Fc.py 

Let me know what u think ?

**Comment by alicera** [comment]
Do you know how to test it ?
for glint360k_r100FC_0.1 , it have the backbone.pt and rank0_softmax_weight.py etc.....
I

**Comment by bokey007** [comment]
check util_callbacks.py file for details.

I think they are using only backbone part here and not the last fully or partial fully connected part.

since during testing u only need to create embedding. no need for last FC layer.

And as they are splitting wt matrix of last layer on eight gpus. U wont need to use any of rankx_softmax_weight.pt or mom.pt
